guidance/_parser.py

```diff
@@ -133,6 +97,54 @@ def _parse(
         return response
 
 
+def process_prompt(prompt_tokens: Sequence[int], ll_interpreter: llguidance.LLInterpreter, bos_token_id: Optional[int]=None) -> list[int]:
+    # Allows ll_interpreter to make adjustments to prompt tokens, such as token healing
+    processed_tokens = ll_interpreter.process_prompt(prompt_tokens)
+    if (
+        bos_token_id is not None
+        and prompt_tokens[:1] != [bos_token_id]
+    ):
+        # add the beginning of sequence token if needed
+        processed_tokens = [bos_token_id] + processed_tokens
+
```

Comment:
These tokens will likely need to be recoded before being sent to the LLM for logits (I *think* only in the case that we just added a BOS token).

You could probably just throw that line in `create_token_parser` after calling `process_prompt`..? Just since you'll have access to a tokenizer there.

See `tests/model_integration/test_model.py::test_associativity`